--- a/interpreter/cling/include/cling/std.modulemap
+++ b/interpreter/cling/include/cling/std.modulemap
@@ -63,6 +63,11 @@ module "std" [system] {
     export *
     header "cmath"
   }
+  module "compare" {
+    requires cplusplus20
+    export *
+    header "compare"
+  }
   module "complex" {
     export *
     header "complex"
@@ -71,6 +76,11 @@ module "std" [system] {
     export *
     header "complex.h"
   }
+  module "concepts" {
+    requires cplusplus20
+    export *
+    header "concepts"
+  }
   module "condition_variable" {
     export *
     header "condition_variable"
@@ -84,7 +94,6 @@ module "std" [system] {
     header "csignal"
   }
   module "cstdalign" {
-    requires !cplusplus17, !header_existence
     export *
     header "cstdalign"
   }
@@ -173,6 +182,7 @@ module "std" [system] {
   }
   module "future" {
     export *
+    export bits_alloc_traits_h
     header "future"
   }
 /*  module "hash_map" {
@@ -244,6 +254,11 @@ module "std" [system] {
     export *
     header "memory"
   }
+  module "memory_resource"  {
+    requires cplusplus17
+    export *
+    header "memory_resource"
+  }
   module "mutex" {
     export *
     header "mutex"
@@ -256,6 +271,11 @@ module "std" [system] {
     export *
     header "numeric"
   }
+  module "optional" {
+    requires cplusplus17
+    export *
+    header "optional"
+  }
   module "ostream" {
     export *
     header "ostream"
@@ -268,12 +288,18 @@ module "std" [system] {
     export *
     header "random"
   }
+  module "ranges" {
+    requires cplusplus20
+    export *
+    header "ranges"
+  }
   module "ratio" {
     export *
     header "ratio"
   }
   module "regex" {
     export *
+    export vector
     header "regex"
   }
   module "scoped_allocator" {
@@ -288,6 +314,11 @@ module "std" [system] {
 //    export *
 //    header "setjmp.h"
 //  }
+  module "span" {
+    requires cplusplus20
+    export *
+    header "span"
+  }
   module "sstream" {
     export *
     header "sstream"
@@ -309,9 +340,10 @@ module "std" [system] {
     header "string"
   }
   module "string_view" {
-    requires !header_existence
+    requires cplusplus17
+
     export *
-    textual header "string_view"
+    header "string_view"
   }
 //  module "string.h" {
 //    export *
@@ -361,24 +393,28 @@ module "std" [system] {
     export *
     header "valarray"
   }
+  module "variant" {
+    requires cplusplus17
+    export *
+    header "variant"
+  }
   module "vector" {
     export *
     header "vector"
   }
   module "codecvt" {
-    requires !header_existence
     export *
     header "codecvt"
   }
   module "cuchar" {
-    requires !header_existence
+    requires cplusplus17
     export *
     header "cuchar"
   }
-  //module "experimental/algorithm" {
-  //  export *
-  //  header "experimental/algorithm"
-  //}
+  module "experimental/string_view" {
+    export *
+    header "experimental/string_view"
+  }
   module "ext/functional" {
     export *
     header "ext/functional"
@@ -396,6 +432,15 @@ module "std" [system] {
     export *
     header "ext/numeric"
   }
+  module "ext/type_traits.h" {
+    export *
+    header "ext/type_traits.h"
+  }
+
+  module "bits_alloc_traits_h" {
+    export *
+    header "bits/alloc_traits.h"
+  }
   module "bits/allocator.h" {
     export *
     header "bits/allocator.h"
@@ -404,6 +449,10 @@ module "std" [system] {
     export *
     header "bits/basic_ios.h"
   }
+  module "bits/chrono.h"  {
+    export *
+    header "bits/chrono.h"
+  }
   module "bits/cpp_type_traits.h" {
     export *
     header "bits/cpp_type_traits.h"
@@ -412,21 +461,113 @@ module "std" [system] {
     export *
     header "bits/exception_defines.h"
   }
+  module "bits_invoke_h" {
+    export *
+    export bits_move_h
+    header "bits/invoke.h"
+  }
   module "bits/ios_base.h" {
     export *
     header "bits/ios_base.h"
   }
+  module "bits/iterator_concepts.h" {
+    requires cplusplus20
+    export *
+    header "bits/iterator_concepts.h"
+  }
   module "bits/locale_facets.h" {
     export *
     header "bits/locale_facets.h"
   }
+  module "bits_move_h" {
+    export *
+    header "bits/move.h"
+  }
+  module "bits_stl_algobase_h" {
+    export *
+    header "bits/stl_algobase.h"
+  }
+  module "bits_stl_iterator_h" {
+    export bits_stl_algobase_h
+    export *
+    header "bits/stl_iterator.h"
+  }
+  module "bits/stl_iterator_base_funcs.h" {
+    export *
+    export bits_stl_iterator_base_types_h
+    header "bits/stl_iterator_base_funcs.h"
+  }
+  module "bits_stl_iterator_base_types_h" {
+    export *
+    export bits_stl_algobase_h
+    header "bits/stl_iterator_base_types.h"
+  }
   module "bits/stl_map.h" {
     export *
     export bits_stl_tree_h
     header "bits/stl_map.h"
   }
+  module "bits/stl_pair.h" {
+    export *
+    header "bits/stl_pair.h"
+  }
   explicit module "bits_stl_tree_h" {
     export *
     header "bits/stl_tree.h"
   }
+  module "bits/range_access.h" {
+    export *
+    export bits_stl_iterator_h
+    header "bits/range_access.h"
+  }
+  module "bits/ranges_algo.h" {
+    requires cplusplus20
+    export *
+    export utility
+    export bits_stl_algo_h
+    header "bits/ranges_algo.h"
+  }
+   module "bits/ranges_algobase.h" {
+    requires cplusplus20
+    export *
+    export initializer_list
+    header "bits/ranges_algobase.h"
+  }
+  explicit module "bits_ranges_base_h"  {
+    requires cplusplus20
+    export *
+    export bits_stl_iterator_h
+    export initializer_list
+    header "bits/ranges_base.h"
+  }
+  module "bits/ranges_util.h"  {
+    requires cplusplus20
+    export *
+    export bits_invoke_h
+    export utility
+    header "bits/ranges_util.h"
+  }
+  module "bits_stl_algo_h" {
+    export *
+    header "bits/stl_algo.h"
+  }
+  module "bits/stl_construct.h" {
+    export *
+    export bits_stl_iterator_base_types_h
+    header "bits/stl_construct.h"
+  }
+  module "bits/uniform_int_dist.h" {
+    export *
+    export bits_stl_algobase_h
+    header "bits/uniform_int_dist.h"
+  }
+  module "bits/utility.h"  {
+    export *
+    header "bits/utility.h"
+  }
+  module "bits/uses_allocator_args.h"  {
+    requires cplusplus17
+    export *
+    header "bits/uses_allocator_args.h"
+  }
 }
