diff --git a/roofit/roofit/inc/RooKeysPdf.h b/roofit/roofit/inc/RooKeysPdf.h
index df0e41232e4..a54f5270c15 100644
--- a/roofit/roofit/inc/RooKeysPdf.h
+++ b/roofit/roofit/inc/RooKeysPdf.h
@@ -1,10 +1,11 @@
 /*****************************************************************************
  * Project: RooFit                                                           *
  * Package: RooFitModels                                                     *
- *    File: $Id: RooKeysPdf.h,v 1.10 2007/05/11 09:13:07 verkerke Exp $
+ *    File: $Id: RooKeysPdf.h,v 1.10 2007/05/11 09:13:07 verkerke Exp $      *
  * Authors:                                                                  *
- *   GR, Gerhard Raven,   UC San Diego,        raven@slac.stanford.edu       *
- *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 * WV, Wouter Verkerke, UC Santa Barbara, verkerke@slac.stanford.edu       *
+ *   GR, Gerhard Raven,   UC San Diego,      raven@slac.stanford.edu         *
+ *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
+ *   WV, Wouter Verkerke, UC Santa Barbara,  verkerke@slac.stanford.edu      *
  *                                                                           *
  * Copyright (c) 2000-2005, Regents of the University of California          *
  *                          and Stanford University. All rights reserved.    *
@@ -16,6 +17,7 @@
 #ifndef ROO_KEYS
 #define ROO_KEYS
 
+#include <vector>
 #include "RooAbsPdf.h"
 #include "RooRealProxy.h"
 
@@ -23,61 +25,68 @@ class RooRealVar;
 
 class RooKeysPdf : public RooAbsPdf {
 public:
-  enum Mirror { NoMirror, MirrorLeft, MirrorRight, MirrorBoth,
-      MirrorAsymLeft, MirrorAsymLeftRight,
-      MirrorAsymRight, MirrorLeftAsymRight,
-      MirrorAsymBoth };
-  RooKeysPdf() ;
-  RooKeysPdf(const char *name, const char *title,
-             RooAbsReal& x, RooDataSet& data, Mirror mirror= NoMirror,
-        double rho=1);
-  RooKeysPdf(const char *name, const char *title,
-             RooAbsReal& x, RooRealVar& xdata, RooDataSet& data, Mirror mirror= NoMirror,
-        double rho=1);
-  RooKeysPdf(const RooKeysPdf& other, const char* name=nullptr);
-  TObject* clone(const char* newname) const override {return new RooKeysPdf(*this,newname); }
-  ~RooKeysPdf() override;
+   enum Mirror {
+      NoMirror,
+      MirrorLeft,
+      MirrorRight,
+      MirrorBoth,
+      MirrorAsymLeft,
+      MirrorAsymLeftRight,
+      MirrorAsymRight,
+      MirrorLeftAsymRight,
+      MirrorAsymBoth
+   };
+   RooKeysPdf();
+   RooKeysPdf(const char *name, const char *title, RooAbsReal &x, RooAbsReal &x0, RooAbsReal &slope, RooAbsReal &offset,
+              RooDataSet &data, Mirror mirror = NoMirror, double rho = 1, bool fast = false, int nPoints = 1000);
+   RooKeysPdf(const char *name, const char *title, RooAbsReal &x, RooRealVar &xdata, RooAbsReal &x0, RooAbsReal &slope,
+              RooAbsReal &offset, RooDataSet &data, Mirror mirror = NoMirror, double rho = 1, bool fast = false,
+              int nPoints = 1000);
+   RooKeysPdf(const RooKeysPdf &other, const char *name = nullptr);
+   TObject *clone(const char *newname) const override { return new RooKeysPdf(*this, newname); }
+   ~RooKeysPdf() override;
 
-  Int_t getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars,
-     const char* rangeName = nullptr) const override;
-  double analyticalIntegral(Int_t code, const char* rangeName = nullptr) const override;
-  Int_t getMaxVal(const RooArgSet& vars) const override;
-  double maxVal(Int_t code) const override;
+   Int_t getAnalyticalIntegral(RooArgSet &allVars, RooArgSet &analVars, const char *rangeName = nullptr) const override;
+   double analyticalIntegral(Int_t code, const char *rangeName = nullptr) const override;
 
-  void LoadDataSet( RooDataSet& data);
+   void LoadDataSet(RooDataSet &data);
 
 protected:
-
-  RooRealProxy _x ;
-  double evaluate() const override;
+   RooRealProxy _x;
+   RooRealProxy _x0;
+   RooRealProxy _slope;
+   RooRealProxy _offset;
+   double evaluate() const override;
 
 private:
-  // how far you have to go out in a Gaussian until it is smaller than the
-  // machine precision
-  static const double _nSigma; //!
+   // how far you have to go out in a Gaussian until it is smaller than the
+   // machine precision
+   const double _nSigma = std::sqrt(-2. * std::log(std::numeric_limits<double>::epsilon()));
 
-  Int_t _nEvents = 0;
-  double *_dataPts = nullptr;  //[_nEvents]
-  double *_dataWgts = nullptr; //[_nEvents]
-  double *_weights = nullptr;  //[_nEvents]
-  double _sumWgt = 0.0;
+   Int_t _nEvents = 0;
+   double *_dataPts = nullptr;  //[_nEvents]
+   double *_dataWgts = nullptr; //[_nEvents]
+   double *_weights = nullptr;  //[_nEvents]
+   double _sumWgt = 0.0;
 
-  constexpr static int _nPoints{1000};
-  double _lookupTable[_nPoints+1];
+   int _nPoints;
+   std::vector<double> _lookupTable;
 
-  double g(double x,double sigma) const;
+   double g(double x, double sigma) const;
+   double interpolate_table(const double &x, const std::vector<double> &table) const;
 
-  bool _mirrorLeft = false;
-  bool _mirrorRight = false;
-  bool _asymLeft = false;
-  bool _asymRight = false;
+   bool _mirrorLeft;
+   bool _mirrorRight;
+   bool _asymLeft;
+   bool _asymRight;
 
-  // cached info on variable
-  Char_t _varName[128];
-  double _lo, _hi, _binWidth;
-  double _rho;
+   // cached info on variable
+   Char_t _varName[128];
+   double _lo, _hi, _binWidth;
+   double _rho;
+   bool _fast;
 
-  ClassDefOverride(RooKeysPdf,2) // One-dimensional non-parametric kernel estimation p.d.f.
+   ClassDefOverride(RooKeysPdf, 2) // One-dimensional non-parametric kernel estimation p.d.f.
 };
 
 #endif
diff --git a/roofit/roofit/src/RooKeysPdf.cxx b/roofit/roofit/src/RooKeysPdf.cxx
index 5361e4b3425..84187b8cf13 100644
--- a/roofit/roofit/src/RooKeysPdf.cxx
+++ b/roofit/roofit/src/RooKeysPdf.cxx
@@ -1,7 +1,7 @@
 /*****************************************************************************
  * Project: RooFit                                                           *
  * Package: RooFitModels                                                     *
- * @(#)root/roofit:$Id$
+ * @(#)root/roofit:$Id$                                                      *
  * Authors:                                                                  *
  *   GR, Gerhard Raven,   UC San Diego,        raven@slac.stanford.edu       *
  *   DK, David Kirkby,    UC Irvine,         dkirkby@uci.edu                 *
@@ -21,7 +21,7 @@
 Class RooKeysPdf implements a one-dimensional kernel estimation p.d.f which model the distribution
 of an arbitrary input dataset as a superposition of Gaussian kernels, one for each data point,
 each contributing 1/N to the total integral of the pdf.
-If the 'adaptive mode' is enabled, the width of the Gaussian is adaptively calculated from the
+The width of the Gaussian is adaptively calculated from the
 local density of events, i.e. narrow for regions with high event density to preserve details and
 wide for regions with low event density to promote smoothness. The details of the general algorithm
 are described in the following paper:
@@ -46,9 +46,6 @@ Cranmer KS, Kernel Estimation in High-Energy Physics.
 
 ClassImp(RooKeysPdf);
 
-const double RooKeysPdf::_nSigma = std::sqrt(-2. *
-    std::log(std::numeric_limits<double>::epsilon()));
-
 ////////////////////////////////////////////////////////////////////////////////
 /// coverity[UNINIT_CTOR]
 
@@ -60,18 +57,25 @@ RooKeysPdf::RooKeysPdf()
 ////////////////////////////////////////////////////////////////////////////////
 /// cache stuff about x
 
-RooKeysPdf::RooKeysPdf(const char *name, const char *title, RooAbsReal &x, RooDataSet &data, Mirror mirror, double rho)
-   : RooKeysPdf(name, title, x, static_cast<RooRealVar &>(x), data, mirror, rho)
+RooKeysPdf::RooKeysPdf(const char *name, const char *title, RooAbsReal &x, RooAbsReal &x0, RooAbsReal &slope,
+                       RooAbsReal &offset, RooDataSet &data, Mirror mirror, double rho, bool fast, int nPoints)
+   : RooKeysPdf(name, title, x, static_cast<RooRealVar &>(x), x0, slope, offset, data, mirror, rho, fast, nPoints)
 {
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 /// cache stuff about x
 
-RooKeysPdf::RooKeysPdf(const char *name, const char *title, RooAbsReal &xpdf, RooRealVar &xdata, RooDataSet &data,
-                       Mirror mirror, double rho)
+RooKeysPdf::RooKeysPdf(const char *name, const char *title, RooAbsReal &xpdf, RooRealVar &xdata, RooAbsReal &x0,
+                       RooAbsReal &slope, RooAbsReal &offset, RooDataSet &data, Mirror mirror, double rho, bool fast,
+                       int nPoints)
    : RooAbsPdf(name, title),
      _x("x", "Observable", this, xpdf),
+     _x0("x0", "Fixed point", this, x0),
+     _slope("slope", "Slope", this, slope),
+     _offset("offset", "Offset", this, offset),
+     _nPoints(nPoints),
+     _lookupTable(_nPoints + 1, 0.),
      _mirrorLeft(mirror == MirrorLeft || mirror == MirrorBoth || mirror == MirrorLeftAsymRight),
      _mirrorRight(mirror == MirrorRight || mirror == MirrorBoth || mirror == MirrorAsymLeftRight),
      _asymLeft(mirror == MirrorAsymLeft || mirror == MirrorAsymLeftRight || mirror == MirrorAsymBoth),
@@ -79,13 +83,14 @@ RooKeysPdf::RooKeysPdf(const char *name, const char *title, RooAbsReal &xpdf, Ro
      _lo(xdata.getMin()),
      _hi(xdata.getMax()),
      _binWidth((_hi - _lo) / (_nPoints - 1)),
-     _rho(rho)
+     _rho(rho),
+     _fast(fast)
 {
-  snprintf(_varName, 128,"%s", xdata.GetName());
+   snprintf(_varName, 128, "%s", xdata.GetName());
 
-  // form the lookup table
-  LoadDataSet(data);
-  TRACE_CREATE;
+   // form the lookup table
+   LoadDataSet(data);
+   TRACE_CREATE;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
@@ -93,7 +98,12 @@ RooKeysPdf::RooKeysPdf(const char *name, const char *title, RooAbsReal &xpdf, Ro
 RooKeysPdf::RooKeysPdf(const RooKeysPdf &other, const char *name)
    : RooAbsPdf(other, name),
      _x("x", this, other._x),
+     _x0("x0", this, other._x0),
+     _slope("slope", this, other._slope),
+     _offset("offset", this, other._offset),
      _nEvents(other._nEvents),
+     _nPoints(other._nPoints),
+     _lookupTable(_nPoints + 1, 0.),
      _mirrorLeft(other._mirrorLeft),
      _mirrorRight(other._mirrorRight),
      _asymLeft(other._asymLeft),
@@ -101,281 +111,328 @@ RooKeysPdf::RooKeysPdf(const RooKeysPdf &other, const char *name)
      _lo(other._lo),
      _hi(other._hi),
      _binWidth(other._binWidth),
-     _rho(other._rho)
+     _rho(other._rho),
+     _fast(other._fast)
 {
-  // cache stuff about x
-  snprintf(_varName, 128, "%s", other._varName );
-
-  // copy over data and weights... not necessary, commented out for speed
-//    _dataPts = new double[_nEvents];
-//    _weights = new double[_nEvents];
-//    for (Int_t i= 0; i<_nEvents; i++) {
-//      _dataPts[i]= other._dataPts[i];
-//      _weights[i]= other._weights[i];
-//    }
-
-  // copy over the lookup table
-  for (Int_t i= 0; i<_nPoints+1; i++)
-    _lookupTable[i]= other._lookupTable[i];
-
-  TRACE_CREATE;
+   // cache stuff about x
+   snprintf(_varName, 128, "%s", other._varName);
+
+   // copy over data and weights... not necessary, commented out for speed
+   //    _dataPts = new double[_nEvents];
+   //    _weights = new double[_nEvents];
+   //    for (Int_t i= 0; i<_nEvents; i++) {
+   //      _dataPts[i]= other._dataPts[i];
+   //      _weights[i]= other._weights[i];
+   //    }
+
+   // copy over the lookup table
+   for (Int_t i = 0; i < _nPoints + 1; i++)
+      _lookupTable[i] = other._lookupTable[i];
+
+   TRACE_CREATE;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
-RooKeysPdf::~RooKeysPdf() {
-  delete[] _dataPts;
-  delete[] _dataWgts;
-  delete[] _weights;
+RooKeysPdf::~RooKeysPdf()
+{
+   delete[] _dataPts;
+   delete[] _dataWgts;
+   delete[] _weights;
 
-  TRACE_DESTROY;
+   TRACE_DESTROY;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 /// small helper structure
 
 namespace {
-  struct Data {
-    double x;
-    double w;
-  };
-  // helper to order two Data structures
-  struct cmp {
-    inline bool operator()(const struct Data& a, const struct Data& b) const
-    { return a.x < b.x; }
-  };
-}
-void RooKeysPdf::LoadDataSet( RooDataSet& data) {
-  delete[] _dataPts;
-  delete[] _dataWgts;
-  delete[] _weights;
-
-  std::vector<Data> tmp;
-  tmp.reserve((1 + _mirrorLeft + _mirrorRight) * data.numEntries());
-  double x0 = 0.;
-  double x1 = 0.;
-  double x2 = 0.;
-  _sumWgt = 0.;
-  // read the data set into tmp and accumulate some statistics
-  RooRealVar& real = static_cast<RooRealVar&>(data.get()->operator[](_varName));
-  for (Int_t i = 0; i < data.numEntries(); ++i) {
-    data.get(i);
-    const double x = real.getVal();
-    const double w = data.weight();
-    x0 += w;
-    x1 += w * x;
-    x2 += w * x * x;
-    _sumWgt += double(1 + _mirrorLeft + _mirrorRight) * w;
-
-    Data p;
-    p.x = x, p.w = w;
-    tmp.push_back(p);
-    if (_mirrorLeft) {
-      p.x = 2. * _lo - x;
-      tmp.push_back(p);
-    }
-    if (_mirrorRight) {
-      p.x = 2. * _hi - x;
+struct Data {
+   double x;
+   double w;
+};
+// helper to order two Data structures
+struct cmp {
+   inline bool operator()(const struct Data &a, const struct Data &b) const { return a.x < b.x; }
+};
+} // namespace
+
+void RooKeysPdf::LoadDataSet(RooDataSet &data)
+{
+   delete[] _dataPts;
+   delete[] _dataWgts;
+   delete[] _weights;
+
+   std::vector<Data> tmp;
+   tmp.reserve((1 + _mirrorLeft + _mirrorRight) * data.numEntries());
+   double x0 = 0.;
+   double x1 = 0.;
+   double x2 = 0.;
+   _sumWgt = 0.;
+   // read the data set into tmp and accumulate some statistics
+   RooRealVar &real = static_cast<RooRealVar &>(data.get()->operator[](_varName));
+   for (Int_t i = 0; i < data.numEntries(); ++i) {
+      data.get(i);
+      const double x = real.getVal();
+      const double w = data.weight();
+      x0 += w;
+      x1 += w * x;
+      x2 += w * x * x;
+      _sumWgt += double(1 + _mirrorLeft + _mirrorRight) * w;
+
+      Data p;
+      p.x = x, p.w = w;
       tmp.push_back(p);
-    }
-  }
-  // sort the entire data set so that values of x are increasing
-  std::sort(tmp.begin(), tmp.end(), cmp());
-
-  // copy the sorted data set to its final destination
-  _nEvents = tmp.size();
-  _dataPts  = new double[_nEvents];
-  _dataWgts = new double[_nEvents];
-  for (unsigned i = 0; i < tmp.size(); ++i) {
-    _dataPts[i] = tmp[i].x;
-    _dataWgts[i] = tmp[i].w;
-  }
-  {
-    // free tmp
-    std::vector<Data> tmp2;
-    tmp2.swap(tmp);
-  }
-
-  double meanv=x1/x0;
-  double sigmav=std::sqrt(x2/x0-meanv*meanv);
-  double h=std::pow(double(4)/double(3),0.2)*std::pow(_sumWgt,-0.2)*_rho;
-  double hmin=h*sigmav*std::sqrt(2.)/10;
-  double norm=h*std::sqrt(sigmav * _sumWgt)/(2.0*std::sqrt(3.0));
-
-  _weights=new double[_nEvents];
-  for(Int_t j=0;j<_nEvents;++j) {
-    _weights[j] = norm / std::sqrt(_dataWgts[j] * g(_dataPts[j],h*sigmav));
-    if (_weights[j]<hmin) _weights[j]=hmin;
-  }
-
-  // The idea below is that beyond nSigma sigma, the value of the exponential
-  // in the Gaussian is well below the machine precision of a double, so it
-  // does not contribute any more. That way, we can limit how many bins of the
-  // binned approximation in _lookupTable we have to touch when filling it.
-  for (Int_t i=0;i<_nPoints+1;++i) _lookupTable[i] = 0.;
-  for(Int_t j=0;j<_nEvents;++j) {
-      const double xlo = std::min(_hi,
-         std::max(_lo, _dataPts[j] - _nSigma * _weights[j]));
-      const double xhi = std::max(_lo,
-         std::min(_hi, _dataPts[j] + _nSigma * _weights[j]));
-      if (xlo >= xhi) continue;
-      const double chi2incr = _binWidth / _weights[j] / std::sqrt(2.);
+      if (_mirrorLeft) {
+         p.x = 2. * _lo - x;
+         tmp.push_back(p);
+      }
+      if (_mirrorRight) {
+         p.x = 2. * _hi - x;
+         tmp.push_back(p);
+      }
+   }
+   // sort the entire data set so that values of x are increasing
+   std::sort(tmp.begin(), tmp.end(), cmp());
+
+   // copy the sorted data set to its final destination
+   _nEvents = tmp.size();
+   _dataPts = new double[_nEvents];
+   _dataWgts = new double[_nEvents];
+   for (unsigned i = 0; i < tmp.size(); ++i) {
+      _dataPts[i] = tmp[i].x;
+      _dataWgts[i] = tmp[i].w;
+   }
+   {
+      // free tmp
+      std::vector<Data> tmp2;
+      tmp2.swap(tmp);
+   }
+
+   const double sqrt2 = std::sqrt(2.);
+   const double sqrt2pi = std::sqrt(2. * TMath::Pi());
+
+   double meanv = x1 / x0;
+   double sigmav = std::sqrt(x2 / x0 - meanv * meanv);
+   double h = std::pow(4. / 3., 0.2) * std::pow(_sumWgt, -0.2) * _rho;
+   double hmin = h * sigmav * sqrt2 / 10;
+   double norm = h * std::sqrt(sigmav * _sumWgt) / (2.0 * std::sqrt(3.0));
+
+   // NOTE There are two differences with respect to the paper:
+   // 1) rho is already included in the h_star definition, so it also affects f0
+   // 2) h_i is further scaled by a factor 2*sqrt(3).
+
+   // Build lookup table for f0 if requested
+   std::vector<double> _lookupTable_f0;
+   if (_fast) {
+      _lookupTable_f0 = std::move(std::vector<double>(_nPoints + 1, 0.));
+      const double h_star = h * sigmav;
+      const double chi2incr = _binWidth / (h_star * sqrt2);
+      for (Int_t j = 0; j < _nEvents; ++j) {
+         const double xlo = std::min(_hi, std::max(_lo, _dataPts[j] - _nSigma * h_star));
+         const double xhi = std::max(_lo, std::min(_hi, _dataPts[j] + _nSigma * h_star));
+         if (xlo >= xhi)
+            continue;
+         const double weightratio = _dataWgts[j] / h_star;
+         const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
+         const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
+         const double x = (double(_nPoints - binlo) * _lo + double(binlo) * _hi) / double(_nPoints);
+         double chi = (x - _dataPts[j]) / (h_star * sqrt2);
+         for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
+            _lookupTable_f0[k] += weightratio * std::exp(-chi * chi);
+         }
+      }
+      if (_asymLeft) {
+         for (Int_t j = 0; j < _nEvents; ++j) {
+            const double xlo = std::min(_hi, std::max(_lo, 2. * _lo - _dataPts[j] + _nSigma * h_star));
+            const double xhi = std::max(_lo, std::min(_hi, 2. * _lo - _dataPts[j] - _nSigma * h_star));
+            if (xlo >= xhi)
+               continue;
+            const double weightratio = _dataWgts[j] / h_star;
+            const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
+            const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
+            const double x = (double(_nPoints - binlo) * _lo + double(binlo) * _hi) / double(_nPoints);
+            double chi = (x - (2. * _lo - _dataPts[j])) / (h_star * sqrt2);
+            for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
+               _lookupTable_f0[k] -= weightratio * std::exp(-chi * chi);
+            }
+         }
+      }
+      if (_asymRight) {
+         for (Int_t j = 0; j < _nEvents; ++j) {
+            const double xlo = std::min(_hi, std::max(_lo, 2. * _hi - _dataPts[j] + _nSigma * h_star));
+            const double xhi = std::max(_lo, std::min(_hi, 2. * _hi - _dataPts[j] - _nSigma * h_star));
+            if (xlo >= xhi)
+               continue;
+            const double weightratio = _dataWgts[j] / h_star;
+            const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
+            const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
+            const double x = (double(_nPoints - binlo) * _lo + double(binlo) * _hi) / double(_nPoints);
+            double chi = (x - (2. * _hi - _dataPts[j])) / (h_star * sqrt2);
+            for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
+               _lookupTable_f0[k] -= weightratio * std::exp(-chi * chi);
+            }
+         }
+      }
+
+      // To match the output of g(), the lookup table for f0 does not include the global 1/n factor.
+      for (Int_t i = 0; i < _nPoints + 1; ++i)
+         _lookupTable_f0[i] /= sqrt2pi;
+   }
+
+   // if fast, use table; If not, use g.
+
+   _weights = new double[_nEvents];
+   for (Int_t j = 0; j < _nEvents; ++j) {
+      // f0 here does not include the global 1/n factor
+      const double f0 = _fast ? interpolate_table(_dataPts[j], _lookupTable_f0) : g(_dataPts[j], h * sigmav);
+      _weights[j] = norm / std::sqrt(_dataWgts[j] * f0);
+      if (_weights[j] < hmin)
+         _weights[j] = hmin;
+   }
+
+   // The idea below is that beyond nSigma sigma, the value of the exponential
+   // in the Gaussian is well below the machine precision of a double, so it
+   // does not contribute any more. That way, we can limit how many bins of the
+   // binned approximation in _lookupTable we have to touch when filling it.
+   for (Int_t j = 0; j < _nEvents; ++j) {
+      const double xlo = std::min(_hi, std::max(_lo, _dataPts[j] - _nSigma * _weights[j]));
+      const double xhi = std::max(_lo, std::min(_hi, _dataPts[j] + _nSigma * _weights[j]));
+      if (xlo >= xhi)
+         continue;
+      const double chi2incr = _binWidth / (_weights[j] * sqrt2);
       const double weightratio = _dataWgts[j] / _weights[j];
       const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
       const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
-      const double x = (double(_nPoints - binlo) * _lo +
-         double(binlo) * _hi) / double(_nPoints);
-      double chi = (x - _dataPts[j]) / _weights[j] / std::sqrt(2.);
+      const double x = (double(_nPoints - binlo) * _lo + double(binlo) * _hi) / double(_nPoints);
+      double chi = (x - _dataPts[j]) / (_weights[j] * sqrt2);
       for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
-     _lookupTable[k] += weightratio * std::exp(- chi * chi);
+         _lookupTable[k] += weightratio * std::exp(-chi * chi);
       }
-  }
-  if (_asymLeft) {
-      for(Int_t j=0;j<_nEvents;++j) {
-     const double xlo = std::min(_hi,
-        std::max(_lo, 2. * _lo - _dataPts[j] + _nSigma * _weights[j]));
-     const double xhi = std::max(_lo,
-        std::min(_hi, 2. * _lo - _dataPts[j] - _nSigma * _weights[j]));
-     if (xlo >= xhi) continue;
-     const double chi2incr = _binWidth / _weights[j] / std::sqrt(2.);
-     const double weightratio = _dataWgts[j] / _weights[j];
-     const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
-     const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
-     const double x = (double(_nPoints - binlo) * _lo +
-        double(binlo) * _hi) / double(_nPoints);
-     double chi = (x - (2. * _lo - _dataPts[j])) / _weights[j] / std::sqrt(2.);
-     for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
-         _lookupTable[k] -= weightratio * std::exp(- chi * chi);
-     }
+   }
+   if (_asymLeft) {
+      for (Int_t j = 0; j < _nEvents; ++j) {
+         const double xlo = std::min(_hi, std::max(_lo, 2. * _lo - _dataPts[j] + _nSigma * _weights[j]));
+         const double xhi = std::max(_lo, std::min(_hi, 2. * _lo - _dataPts[j] - _nSigma * _weights[j]));
+         if (xlo >= xhi)
+            continue;
+         const double chi2incr = _binWidth / (_weights[j] * sqrt2);
+         const double weightratio = _dataWgts[j] / _weights[j];
+         const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
+         const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
+         const double x = (double(_nPoints - binlo) * _lo + double(binlo) * _hi) / double(_nPoints);
+         double chi = (x - (2. * _lo - _dataPts[j])) / (_weights[j] * sqrt2);
+         for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
+            _lookupTable[k] -= weightratio * std::exp(-chi * chi);
+         }
       }
-  }
-  if (_asymRight) {
-      for(Int_t j=0;j<_nEvents;++j) {
-     const double xlo = std::min(_hi,
-        std::max(_lo, 2. * _hi - _dataPts[j] + _nSigma * _weights[j]));
-     const double xhi = std::max(_lo,
-        std::min(_hi, 2. * _hi - _dataPts[j] - _nSigma * _weights[j]));
-     if (xlo >= xhi) continue;
-     const double chi2incr = _binWidth / _weights[j] / std::sqrt(2.);
-     const double weightratio = _dataWgts[j] / _weights[j];
-     const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
-     const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
-     const double x = (double(_nPoints - binlo) * _lo +
-        double(binlo) * _hi) / double(_nPoints);
-     double chi = (x - (2. * _hi - _dataPts[j])) / _weights[j] / std::sqrt(2.);
-     for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
-         _lookupTable[k] -= weightratio * std::exp(- chi * chi);
-     }
+   }
+   if (_asymRight) {
+      for (Int_t j = 0; j < _nEvents; ++j) {
+         const double xlo = std::min(_hi, std::max(_lo, 2. * _hi - _dataPts[j] + _nSigma * _weights[j]));
+         const double xhi = std::max(_lo, std::min(_hi, 2. * _hi - _dataPts[j] - _nSigma * _weights[j]));
+         if (xlo >= xhi)
+            continue;
+         const double chi2incr = _binWidth / (_weights[j] * sqrt2);
+         const double weightratio = _dataWgts[j] / _weights[j];
+         const Int_t binlo = static_cast<Int_t>(std::floor((xlo - _lo) / _binWidth));
+         const Int_t binhi = static_cast<Int_t>(_nPoints - std::floor((_hi - xhi) / _binWidth));
+         const double x = (double(_nPoints - binlo) * _lo + double(binlo) * _hi) / double(_nPoints);
+         double chi = (x - (2. * _hi - _dataPts[j])) / (_weights[j] * sqrt2);
+         for (Int_t k = binlo; k <= binhi; ++k, chi += chi2incr) {
+            _lookupTable[k] -= weightratio * std::exp(-chi * chi);
+         }
       }
-  }
-  static const double sqrt2pi(std::sqrt(2*TMath::Pi()));
-  for (Int_t i=0;i<_nPoints+1;++i)
-    _lookupTable[i] /= sqrt2pi * _sumWgt;
+   }
+
+   for (Int_t i = 0; i < _nPoints + 1; ++i)
+      _lookupTable[i] /= sqrt2pi * _sumWgt;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
-double RooKeysPdf::evaluate() const {
-  Int_t i = (Int_t)floor((double(_x)-_lo)/_binWidth);
-  if (i<0) {
-//     cerr << "got point below lower bound:"
-//     << double(_x) << " < " << _lo
-//     << " -- performing linear extrapolation..." << endl;
-    i=0;
-  }
-  if (i>_nPoints-1) {
-//     cerr << "got point above upper bound:"
-//     << double(_x) << " > " << _hi
-//     << " -- performing linear extrapolation..." << endl;
-    i=_nPoints-1;
-  }
-  double dx = (double(_x)-(_lo+i*_binWidth))/_binWidth;
-
-  // for now do simple linear interpolation.
-  // one day replace by splines...
-  double ret = (_lookupTable[i]+dx*(_lookupTable[i+1]-_lookupTable[i]));
-  if (ret<0) ret=0 ;
-  return ret ;
-}
-
-Int_t RooKeysPdf::getAnalyticalIntegral(
-   RooArgSet& allVars, RooArgSet& analVars, const char* /* rangeName */) const
+double RooKeysPdf::interpolate_table(const double &x, const std::vector<double> &table) const
 {
-  if (matchArgs(allVars, analVars, _x)) return 1;
-  return 0;
+   Int_t i = (Int_t)floor((x - _lo) / _binWidth);
+   if (i < 0) {
+      i = 0;
+   }
+   if (i > _nPoints - 1) {
+      i = _nPoints - 1;
+   }
+   const double dx = (x - (_lo + i * _binWidth)) / _binWidth;
+   return table[i] + dx * (table[i + 1] - table[i]);
 }
 
-double RooKeysPdf::analyticalIntegral(Int_t code, const char* rangeName) const
+double RooKeysPdf::evaluate() const
 {
-  R__ASSERT(1 == code);
-  // this code is based on _lookupTable and uses linear interpolation, just as
-  // evaluate(); integration is done using the trapez rule
-  const double xmin = std::max(_lo, _x.min(rangeName));
-  const double xmax = std::min(_hi, _x.max(rangeName));
-  const Int_t imin = (Int_t)floor((xmin - _lo) / _binWidth);
-  const Int_t imax = std::min((Int_t)floor((xmax - _lo) / _binWidth),
-      _nPoints - 1);
-  double sum = 0.;
-  // sum up complete bins in middle
-  if (imin + 1 < imax)
-    sum += _lookupTable[imin + 1] + _lookupTable[imax];
-  for (Int_t i = imin + 2; i < imax; ++i)
-    sum += 2. * _lookupTable[i];
-  sum *= _binWidth * 0.5;
-  // treat incomplete bins
-  const double dxmin = (xmin - (_lo + imin * _binWidth)) / _binWidth;
-  const double dxmax = (xmax - (_lo + imax * _binWidth)) / _binWidth;
-  if (imin < imax) {
-    // first bin
-    sum += _binWidth * (1. - dxmin) * 0.5 * (_lookupTable[imin + 1] +
-   _lookupTable[imin] + dxmin *
-   (_lookupTable[imin + 1] - _lookupTable[imin]));
-    // last bin
-    sum += _binWidth * dxmax * 0.5 * (_lookupTable[imax] +
-   _lookupTable[imax] + dxmax *
-   (_lookupTable[imax + 1] - _lookupTable[imax]));
-  } else if (imin == imax) {
-    // first bin == last bin
-    sum += _binWidth * (dxmax - dxmin) * 0.5 * (
-   _lookupTable[imin] + dxmin *
-   (_lookupTable[imin + 1] - _lookupTable[imin]) +
-   _lookupTable[imax] + dxmax *
-   (_lookupTable[imax + 1] - _lookupTable[imax]));
-  }
-  return sum;
+   const double x = (_x - _x0 - _offset) / _slope + _x0;
+   double ret = interpolate_table(x, _lookupTable);
+   if (ret < 0)
+      ret = 0;
+   return ret;
 }
 
-Int_t RooKeysPdf::getMaxVal(const RooArgSet& vars) const
+Int_t RooKeysPdf::getAnalyticalIntegral(RooArgSet &allVars, RooArgSet &analVars, const char * /* rangeName */) const
 {
-  if (vars.contains(*_x.absArg())) return 1;
-  return 0;
+   if (matchArgs(allVars, analVars, _x))
+      return 1;
+   return 0;
 }
 
-double RooKeysPdf::maxVal(Int_t code) const
+double RooKeysPdf::analyticalIntegral(Int_t code, const char *rangeName) const
 {
-  R__ASSERT(1 == code);
-  double max = -std::numeric_limits<double>::max();
-  for (Int_t i = 0; i <= _nPoints; ++i)
-    if (max < _lookupTable[i]) max = _lookupTable[i];
-  return max;
+   R__ASSERT(1 == code);
+   // this code is based on _lookupTable and uses linear interpolation, just as
+   // evaluate(); integration is done using the trapez rule
+   const double rmin = (_x.min(rangeName) - _x0 - _offset) / _slope + _x0;
+   const double rmax = (_x.max(rangeName) - _x0 - _offset) / _slope + _x0;
+   const double xmin = std::max(_lo, rmin);
+   const double xmax = std::min(_hi, rmax);
+   const Int_t imin = (Int_t)floor((xmin - _lo) / _binWidth);
+   const Int_t imax = std::min((Int_t)floor((xmax - _lo) / _binWidth), _nPoints - 1);
+   double sum = 0.;
+   // sum up complete bins in middle
+   if (imin + 1 < imax)
+      sum += _lookupTable[imin + 1] + _lookupTable[imax];
+   for (Int_t i = imin + 2; i < imax; ++i)
+      sum += 2. * _lookupTable[i];
+   sum *= _binWidth * 0.5;
+   // treat incomplete bins
+   const double dxmin = (xmin - (_lo + imin * _binWidth)) / _binWidth;
+   const double dxmax = (xmax - (_lo + imax * _binWidth)) / _binWidth;
+   if (imin < imax) {
+      // first bin
+      sum += _binWidth * (1. - dxmin) * 0.5 *
+             (_lookupTable[imin + 1] + _lookupTable[imin] + dxmin * (_lookupTable[imin + 1] - _lookupTable[imin]));
+      // last bin
+      sum += _binWidth * dxmax * 0.5 *
+             (_lookupTable[imax] + _lookupTable[imax] + dxmax * (_lookupTable[imax + 1] - _lookupTable[imax]));
+   } else if (imin == imax) {
+      // first bin == last bin
+      sum += _binWidth * (dxmax - dxmin) * 0.5 *
+             (_lookupTable[imin] + dxmin * (_lookupTable[imin + 1] - _lookupTable[imin]) + _lookupTable[imax] +
+              dxmax * (_lookupTable[imax + 1] - _lookupTable[imax]));
+   }
+   return sum * _slope;
 }
 
 ////////////////////////////////////////////////////////////////////////////////
 
-double RooKeysPdf::g(double x,double sigmav) const {
-  double y=0;
-  // since data is sorted, we can be a little faster because we know which data
-  // points contribute
-  double* it = std::lower_bound(_dataPts, _dataPts + _nEvents,
-      x - _nSigma * sigmav);
-  if (it >= (_dataPts + _nEvents)) return 0.;
-  double* iend = std::upper_bound(it, _dataPts + _nEvents,
-      x + _nSigma * sigmav);
-  for ( ; it < iend; ++it) {
-    const double r = (x - *it) / sigmav;
-    y += std::exp(-0.5 * r * r);
-  }
-
-  static const double sqrt2pi(std::sqrt(2*TMath::Pi()));
-  return y/(sigmav*sqrt2pi);
+double RooKeysPdf::g(double x, double sigmav) const
+{
+   double y = 0;
+   // since data is sorted, we can be a little faster because we know which data
+   // points contribute
+   double *it = std::lower_bound(_dataPts, _dataPts + _nEvents, x - _nSigma * sigmav);
+   if (it >= (_dataPts + _nEvents))
+      return 0.;
+   double *iend = std::upper_bound(it, _dataPts + _nEvents, x + _nSigma * sigmav);
+   for (; it < iend; ++it) {
+      const double r = (x - *it) / sigmav;
+      y += std::exp(-0.5 * r * r);
+   }
+
+   static const double sqrt2pi(std::sqrt(2 * TMath::Pi()));
+   return y / (sigmav * sqrt2pi);
 }
