diff --git a/roofit/batchcompute/res/RooBatchCompute.h b/roofit/batchcompute/res/RooBatchCompute.h
index 7b79907d5ce..a6431c0d066 100644
--- a/roofit/batchcompute/res/RooBatchCompute.h
+++ b/roofit/batchcompute/res/RooBatchCompute.h
@@ -94,6 +94,7 @@ enum Computer {
    Poisson,
    Polynomial,
    Power,
+   PowerLaw,
    ProdPdf,
    Ratio,
    TruthModelExpBasis,
diff --git a/roofit/batchcompute/src/ComputeFunctions.cxx b/roofit/batchcompute/src/ComputeFunctions.cxx
index 78e8ca3a7d2..2045605d3dc 100644
--- a/roofit/batchcompute/src/ComputeFunctions.cxx
+++ b/roofit/batchcompute/src/ComputeFunctions.cxx
@@ -766,6 +766,16 @@ __rooglobal__ void computePower(Batches &batches)
    }
 }
 
+__rooglobal__ void computePowerLaw(Batches &batches)
+{
+   Batch x = batches.args[0];
+   Batch x0 = batches.args[1];
+   Batch c = batches.args[2];
+   for (size_t i = BEGIN; i < batches.nEvents; i += STEP) {
+      batches.output[i] = (x[i] > x0[i]) ? std::pow(x[i] - x0[i], c[i]) : 0.;
+   }
+}
+
 __rooglobal__ void computeProdPdf(Batches &batches)
 {
    const int nPdfs = batches.extra[0];
@@ -949,6 +959,7 @@ std::vector<void (*)(Batches &)> getFunctions()
            computePoisson,
            computePolynomial,
            computePower,
+           computePowerLaw,
            computeProdPdf,
            computeRatio,
            computeTruthModelExpBasis,
diff --git a/roofit/roofit/CMakeLists.txt b/roofit/roofit/CMakeLists.txt
index 0e653a47a85..8213aedbb32 100644
--- a/roofit/roofit/CMakeLists.txt
+++ b/roofit/roofit/CMakeLists.txt
@@ -66,6 +66,7 @@ ROOT_STANDARD_LIBRARY_PACKAGE(RooFit
     RooParamHistFunc.h
     RooPoisson.h
     RooPolynomial.h
+    RooPowerLaw.h
     RooSpline.h
     RooStepFunction.h
     RooTFnBinding.h
@@ -129,6 +130,7 @@ ROOT_STANDARD_LIBRARY_PACKAGE(RooFit
     src/RooParamHistFunc.cxx
     src/RooPoisson.cxx
     src/RooPolynomial.cxx
+    src/RooPowerLaw.cxx
     src/RooSpline.cxx
     src/RooStepFunction.cxx
     src/RooTFnBinding.cxx
diff --git a/roofit/roofit/inc/LinkDef1.h b/roofit/roofit/inc/LinkDef1.h
index 06d3fb51d93..10cb1dd6186 100644
--- a/roofit/roofit/inc/LinkDef1.h
+++ b/roofit/roofit/inc/LinkDef1.h
@@ -35,6 +35,7 @@
 #pragma link C++ class RooParamHistFunc+ ;
 #pragma link C++ class RooParametricStepFunction+ ;
 #pragma link C++ class RooPolynomial+ ;
+#pragma link C++ class RooPowerLaw+ ;
 #pragma link C++ class RooUnblindCPAsymVar+ ;
 #pragma link C++ class RooUnblindOffset+ ;
 #pragma link C++ class RooUnblindPrecision+ ;
diff --git a/roofit/roofit/inc/RooPowerLaw.h b/roofit/roofit/inc/RooPowerLaw.h
new file mode 100644
index 00000000000..f249f68b29f
--- /dev/null
+++ b/roofit/roofit/inc/RooPowerLaw.h
@@ -0,0 +1,51 @@
+/*****************************************************************************
+ * Project: RooFit                                                           *
+ * Package: RooFitModels                                                     *
+ *                                                                           *
+ * Copyright (c) 2000-2005, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+#ifndef ROO_POWERLAW
+#define ROO_POWERLAW
+
+#include "RooAbsPdf.h"
+#include "RooRealProxy.h"
+
+class RooPowerLaw : public RooAbsPdf {
+public:
+  RooPowerLaw() {} ;
+  RooPowerLaw(const char *name, const char *title,
+       RooAbsReal& _x, RooAbsReal& _x0, RooAbsReal& _c);
+  RooPowerLaw(const RooPowerLaw& other, const char* name = nullptr);
+  TObject* clone(const char* newname) const override { return new RooPowerLaw(*this, newname); }
+
+  /// Get the x variable.
+  RooAbsReal const& getX() const { return x.arg(); }
+
+  /// Get the coefficient "c0".
+  RooAbsReal const& getExponent() const { return c.arg(); }
+
+  /// Get the coefficient "x0".
+  RooAbsReal const &getShift() const { return x0.arg(); }
+
+  Int_t getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars, const char* rangeName = nullptr) const override;
+  double analyticalIntegral(Int_t code, const char* rangeName = nullptr) const override;
+
+protected:
+  RooRealProxy x;
+  RooRealProxy x0;
+  RooRealProxy c;
+
+  double evaluate() const override;
+  void doEval(RooFit::EvalContext &) const override;
+  inline bool canComputeBatchWithCuda() const override { return true; }
+
+private:
+  ClassDefOverride(RooPowerLaw, 1) // PowerLaw PDF
+};
+
+#endif
diff --git a/roofit/roofit/src/RooPowerLaw.cxx b/roofit/roofit/src/RooPowerLaw.cxx
new file mode 100644
index 00000000000..590ee619f79
--- /dev/null
+++ b/roofit/roofit/src/RooPowerLaw.cxx
@@ -0,0 +1,113 @@
+/*****************************************************************************
+ * Project: RooFit                                                           *
+ * Package: RooFitModels                                                     *
+ * @(#)root/roofit:$Id$
+ *                                                                           *
+ * Copyright (c) 2000-2005, Regents of the University of California          *
+ *                          and Stanford University. All rights reserved.    *
+ *                                                                           *
+ * Redistribution and use in source and binary forms,                        *
+ * with or without modification, are permitted according to the terms        *
+ * listed in LICENSE (http://roofit.sourceforge.net/license.txt)             *
+ *****************************************************************************/
+
+/** \class RooPowerLaw
+    \ingroup Roofit
+
+Power law PDF. It computes
+\f[
+  \mathrm{RooPowerLaw}(x, x_0, c) = \mathcal{N} \cdot (x - x_0)^c,
+\f]
+where \f$ \mathcal{N} \f$ is a normalization constant that depends on the
+range and values of the arguments.
+
+Note: Adapted from RooExponential for RDX run 2 analysis.
+**/
+
+#include "RooPowerLaw.h"
+
+#include "RooRealVar.h"
+#include "RooBatchCompute.h"
+
+#include "TError.h"
+
+#include <cmath>
+
+////////////////////////////////////////////////////////////////////////////////
+
+RooPowerLaw::RooPowerLaw(const char *name, const char *title,
+                RooAbsReal& _x, RooAbsReal& _x0, RooAbsReal& _c) :
+  RooAbsPdf(name, title),
+  x("x", "Dependent", this, _x),
+  x0("x0", "Shift", this, _x0),
+  c("c", "Exponent", this, _c)
+{
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+RooPowerLaw::RooPowerLaw(const RooPowerLaw& other, const char* name) :
+  RooAbsPdf(other, name),
+  x("x", this, other.x),
+  x0("x0", this, other.x0),
+  c("c", this, other.c)
+{
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+double RooPowerLaw::evaluate() const
+{
+  if (x > x0) {
+    return std::pow(x - x0, c);
+  } else {
+    return 0.;
+  }
+}
+
+////////////////////////////////////////////////////////////////////////////////
+/// Compute multiple values of Exponential distribution.
+void RooPowerLaw::doEval(RooFit::EvalContext &ctx) const
+{
+   auto computer = RooBatchCompute::Exponential;
+   RooBatchCompute::compute(ctx.config(this), computer, ctx.output(), {ctx.at(x), ctx.at(c)});
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+Int_t RooPowerLaw::getAnalyticalIntegral(RooArgSet& allVars, RooArgSet& analVars, const char* /*rangeName*/) const
+{
+  if (matchArgs(allVars,analVars,x)) return 1;
+  return 0 ;
+}
+
+////////////////////////////////////////////////////////////////////////////////
+
+double RooPowerLaw::analyticalIntegral(Int_t code, const char* rangeName) const
+{
+  R__ASSERT(code == 1);
+
+  auto& constant  = c;
+  auto& integrand = x;
+
+  double max = integrand.max(rangeName);
+  double min = integrand.min(rangeName);
+
+  // If the whole integration region is below threshold, return 0
+  if (max <= x0) {
+    return 0.;
+  }
+
+  // If only part of the integration region is below threshold, integrate from x0 to max
+  if (min < x0) {
+    min = x0;
+  }
+
+  if (constant == -1.0) {
+    return std::log(max - x0) - std::log(min - x0);
+  }
+
+  auto k = constant + 1.;
+
+  return (std::pow(max - x0, k) - std::pow(min - x0, k)) / k;
+}
