--- a/io/io/src/TBufferFile.cxx
+++ b/io/io/src/TBufferFile.cxx
@@ -48,7 +48,7 @@
 const UInt_t kNewClassTag       = 0xFFFFFFFF;
 const UInt_t kClassMask         = 0x80000000;  // OR the class index with this
 const UInt_t kByteCountMask     = 0x40000000;  // OR the byte count with this
-const UInt_t kMaxMapCount       = 0x3FFFFFFE;  // last valid fMapCount and byte count
+const UInt_t kMaxMapCount       = 0x9FFFFFFE;  // last valid fMapCount and byte count
 const Version_t kByteCountVMask = 0x4000;      // OR the version byte count with this
 const Version_t kMaxVersion     = 0x3FFF;      // highest possible version number
 const Int_t  kMapOffset         = 2;   // first 2 map entries are taken by null obj and self obj