diff --git a/config/root-config.in b/config/root-config.in
index 193551e601..4494a18bf5 100755
--- a/config/root-config.in
+++ b/config/root-config.in
@@ -276,42 +276,17 @@ openbsd)
    ;;
 macosx)
    # MacOS X with gcc (GNU cc v3.1) and possible fink (fink.sf.net)
-   macosx_minor=`sw_vers | sed -n 's/ProductVersion://p' | cut -d . -f 2`
-   # cannot find the one linked to libGraf if relocated after built
-   if [ $macosx_minor -le 4 ]; then
-      rootlibs="$rootlibs -lfreetype"
-   fi
-   if [ $macosx_minor -le 3 ]; then
-      finkdir=`which fink 2>&1 | sed -ne "s/\/bin\/fink//p"`
-      auxcflags=`[ -d ${finkdir}/include ] && echo -I${finkdir}/include`
-      auxcflags="-Wno-long-double $auxcflags"
-      auxlibs="-lm `[ -d ${finkdir}/lib ] && echo -L${finkdir}/lib` -ldl"
-      forcelibs=$rootulibs
-      forceglibs=$rootuglibs
-      forceevelibs=$rootuevelibs
-   else
-      auxcflags="${cxxversionflag} -m32"
-      auxldflags="-m32"
-      auxlibs="-lm -ldl"
-   fi
+  auxcflags="${cxxversionflag} -m32"
+  auxldflags="-m32"
+  auxlibs="-lm -ldl"
    ;;
 macosxicc)
    # MacOS X with Intel icc compiler
-   macosx_minor=`sw_vers | sed -n 's/ProductVersion://p' | cut -d . -f 2`
-   # cannot find the one linked to libGraf if relocated after built
-   if [ $macosx_minor -le 4 ]; then
-      rootlibs="$rootlibs -lfreetype"
-   fi
    auxcflags="${cxxversionflag}"
    auxlibs="-lm -ldl"
    ;;
 macosx64)
    # MacOS X with gcc (GNU cc v4.x) in 64 bit mode
-   macosx_minor=`sw_vers | sed -n 's/ProductVersion://p' | cut -d . -f 2`
-   # cannot find the one linked to libGraf if relocated after built
-   if [ $macosx_minor -le 4 ]; then
-      rootlibs="$rootlibs -lfreetype"
-   fi
    auxcflags="${cxxversionflag} -m64"
    auxldflags="-m64"
    auxlibs="-lm -ldl"
@@ -375,18 +350,11 @@ freebsd* | openbsd* | linux*)
 macosx*)
    for f in $features ; do
       if test "x$f" = "xthread" ; then
-         if [ $macosx_minor -ge 5 ]; then
             auxcflags="-pthread $auxcflags"
             auxlibs="-lpthread $auxlibs"
-         else
-            auxcflags="-D_REENTRANT $auxcflags"
-            auxlibs="-lpthread $auxlibs"
-         fi
       fi
       if test "x$f" = "xrpath" ; then
-         if [ $macosx_minor -ge 5 ]; then
             auxlibs="-Wl,-rpath,$libdir $auxlibs"
-         fi
       fi
       if test "x$f" = "xlibcxx" ; then
          auxcflags="-stdlib=libc++ $auxcflags"
