diff --git a/hist/hist/src/TH1.cxx b/hist/hist/src/TH1.cxx
index 80f3e527e5..147092a941 100644
--- a/hist/hist/src/TH1.cxx
+++ b/hist/hist/src/TH1.cxx
@@ -5391,9 +5391,9 @@ static inline bool IsEquidistantBinning(const TAxis& axis)
 
 Bool_t TH1::SameLimitsAndNBins(const TAxis& axis1, const TAxis& axis2)
 {
-   return axis1.GetNbins() == axis2.GetNbins()
-      && axis1.GetXmin() == axis2.GetXmin()
-      && axis1.GetXmax() == axis2.GetXmax();
+  return axis1.GetNbins() == axis2.GetNbins() &&
+      TMath::AreEqualAbs(axis1.GetXmin(), axis2.GetXmin(), axis1.GetBinWidth(axis1.GetNbins()) * 1.E-7) &&
+      TMath::AreEqualAbs(axis1.GetXmax(), axis2.GetXmax(), axis1.GetBinWidth(axis1.GetNbins()) * 1.E-7);
 }
 
 ////////////////////////////////////////////////////////////////////////////////
