diff --git a/include/Hammer/Math/Units.hh b/include/Hammer/Math/Units.hh
index 91a720be..df51b494 100644
--- a/include/Hammer/Math/Units.hh
+++ b/include/Hammer/Math/Units.hh
@@ -13,6 +13,7 @@
 #define HAMMER_MATH_UNITS
 
 #include <map>
+#include <string>
 
 namespace Hammer {
 
diff --git a/include/Hammer/Math/Utils.hh b/include/Hammer/Math/Utils.hh
index bc3cb11f..023f2473 100644
--- a/include/Hammer/Math/Utils.hh
+++ b/include/Hammer/Math/Utils.hh
@@ -14,6 +14,8 @@
 
 #include <complex>
 #include <limits>
+#include <cstring>
+#include <climits>
 
 namespace Hammer {
 
